packages/runtime/src/state/modules/element-trees.ts

```diff
@@ -0,0 +1,358 @@
+  updateParentElements(elements, parentElementPaths, rootElements.new)
+
+  return {
+    elements,
+    elementIds,
+  }
+}
+
+function insertElement(
+  { elements, elementIds }: ElementTree,
+  insertedElement: Element,
+  propName: string,
+  descriptors: DescriptorsByComponentType,
+) {
+  if (hasChildren(insertedElement, propName, descriptors)) {
+    for (const element of traverseElementTree(insertedElement, descriptors)) {
+      elements.set(element.key, element)
+      if (!isElementReference(element)) {
+        const elementId = getElementIdProp(element, descriptors)
+        if (elementId != null) elementIds.set(element.key, elementId)
+      }
+    }
+  } else {
+    elements.set(insertedElement.key, insertedElement)
+    if (!isElementReference(insertedElement)) {
+      const elementId = getElementIdProp(insertedElement, descriptors)
+      if (elementId != null) elementIds.set(insertedElement.key, elementId)
+    }
+  }
+}
+
+function applyInsert(
+  elementTree: ElementTree,
+  descriptors: DescriptorsByComponentType,
+  rootElements: { old: Element; new: Element },
+  path: OperationPath,
+): ElementTree {
+  const [insertedElementPath, ...parentElementPaths] = getChangedElementsPaths(path)
+  const [insertedElement, propName] = getElementAndPropName(rootElements.new, insertedElementPath)
+
+  const elements = new Map(elementTree.elements)
+  const elementIds = new Map(elementTree.elementIds)
+
+  insertElement({ elements, elementIds }, insertedElement, propName, descriptors)
+  updateParentElements(elements, parentElementPaths, rootElements.new)
+
+  return {
+    elements,
+    elementIds,
+  }
+}
+
+function applyUpdate(
+  elementTree: ElementTree,
+  descriptors: DescriptorsByComponentType,
+  rootElements: { old: Element; new: Element },
+  path: OperationPath,
+): ElementTree {
+  const [updateElementPath, ...parentElementPaths] = getChangedElementsPaths(path)
+  const [deletedElement, propName] = getElementAndPropName(rootElements.old, updateElementPath)
+  const [insertedElement, _] = getElementAndPropName(rootElements.new, updateElementPath)
+
+  const elements = new Map(elementTree.elements)
+  const elementIds = new Map(elementTree.elementIds)
+
+  deleteElement({ elements, elementIds }, deletedElement, propName, descriptors)
+  insertElement({ elements, elementIds }, insertedElement, propName, descriptors)
+
+  updateParentElements(elements, parentElementPaths, rootElements.new)
+
+  return {
+    elements,
+    elementIds,
+  }
+}
+
+function applyChanges(
+  elementTree: ElementTree,
+  descriptors: DescriptorsByComponentType,
+  rootElements: { old: Element; new: Element },
+  operation: Operation,
+): ElementTree {
+  return operation.reduce((tree, op) => {
+    const hasDelete = 'ld' in op || 'od' in op
+    const hasInsert = 'li' in op || 'oi' in op
+    if (hasDelete && hasInsert) {
+      return applyUpdate(tree, descriptors, rootElements, op.p)
```

Comment:
Technically, an update can be just handled as a delete followed by an insert; special casing to avoid doubly updating parent elements.